### DISCUSSION TITLE
Why semicolons?

### DISCUSSION THREAD

**Discussion by yerlaser** [c3lang/c3c#1740] [General]
Overall seems like a sound idea to create a better and simpler C.
But, why mandatory semicolons?
They looks so-o outdated.

**Comment by lerno** [comment]
It reduces ambiguity. It reduces pain going to C and back.

**Reply by yerlaser** [reply]
That's valid, but only for those who need to go back and forth between C and C3.
But, for newcomers (especially those coming from JS or Python) it's a pain.
With regard to ambiguity, it has been shown many times that semicolons don't improve on it.
They instead add clutter to the code.

It's a relict of the very early compilers who couldn't handle this, but I don't understand why one would include this in a modern language.

**Reply by lerno** [reply]
> With regard to ambiguity, it has been shown many times that semicolons don't improve on it.

By whom? The grammar now needs to be whitespace sensitive and in order to allow useful multi-line constructs, special rules must be added. See for example https://pling.jondgoodwin.com/post/semicolon-inference/ for an overview.

> It's a relict of the very early compilers who couldn't handle this, but I don't understand why one would include this in a modern language.

BASIC didn't use an EOS marker and it's not in any way a "modern" language.

The reason you would include EOS markers are:

1. Better error messages
2. Straightforward grammar
3. Simpler tooling
4. Simpler syntax highlighting
5. Follows C syntax

As stated here: https://c3-lang.org/getting-started/design-goals/

> Stay close to C - only change where there is a significant need.

Is there a *need* to change to not using semicolons?

The motivation for not using semicolons are apparently:

1. It's not "modern"
2. It's not easy for people coming from JS or Python

Designing a language on what's currently perceived as "modern" is possible. That is a non-goal for C3 though.

For a Python or JS programmer there will be other more significant obstacles, such as understanding how to work in a memory managed language. And why should *those* programmers be especially catered to? Why not C, C++, Java or PHP programmers?

Why should programmers from non-semi languages be catered to, but not programmers from other languages?

In the end, to change from C syntax, extraordinary arguments are needed. Catering to what Python programmers are used to (request for indentation based scopes when?) and that it might not feel "modern" are not extraordinary arguments.

**Reply by yerlaser** [reply]
These semicolons imho are like wearing a tie, at least as it seems to me.
Nobody can explain exactly what practical advantage they provide.
At the same time many other popular languages got rid of them proving that it's not a problem for a modern compiler.
And no, you don't need to make whitespace significant for that, but, yes, some other rules are needed.
This is a one time investment from both the compiler and the users and this relict could go away.

But, anyway, I didn't mean to annoy anyone.
Wish the project good luck anyway!

**Reply by lerno** [reply]
Semicolons has a very practical use in making the grammar unambiguous. *All* languages that do not use EOS markers (there are languages with other end-of-statement markers, like Smalltalk uses "." for example) need to have some heuristics or special rules to allow the very common case where a statement needs to span more than one line.

Again, feel free to read the linked article, or any of the many other articles on the subject. For example, JS fails without semicolons in the following example:

```js
const a = 5
const b = 10
const sum = a + b
(function() {
  console.log(sum)
})()
```

Is it a practical advantage to not have to worry about when semicolons are needed and when they aren't? I would say it is.

And as previously mentioned, this also affects tooling, which can be simpler when there is an EOS marker.
